Add building Docker image for aarch64

fisuda opened the pull request:
This PR adds aarch64 support. The Orion can run in a docker container on Raspberry Pi by this.

I have run the functional test and the valgrind test, and created the coverage report. The test logs are here (https://github.com/fisuda/report/tree/master/orion/20200225_orion_aarch64).

In the function test, two tests have failed as shown:

- 1585_log_summary
- 1697_more_than_one_bad_input

It seems that the issue is dependent on the timing. So, they passed by changing the value of `-logSummary` option.

It would be great if you could review this PR.

Thanks!

fgalan commented on doc/manuals/admin/build_source.md:
```diff
@@ -53,6 +53,7 @@ commands that require root privilege):
 
 * Install Google Test/Mock from sources (there are RPM packages for this, but they do not work with the current CMakeLists.txt configuration). Previously the URL was http://googlemock.googlecode.com/files/gmock-1.5.0.tar.bz2 but Google removed that package in late August 2016 and it is no longer working.
 
+        yum install perl-Digest-MD5 libxslt
```
I understand this refers to the standard procedure to build Orion from soures (i.e. not specific for aarch64).b Could you elaborate on the need of these packages, please?

fisuda replied:
In the case of aarch64, as the following errors occurred in coverage, they are needed.
- perl-Digest-MD5
```
make[1]: Leaving directory `/opt/fiware-orion/BUILD_COVERAGE'
# Init coverage
echo "Initializing coverage files"
Initializing coverage files
mkdir -p coverage
lcov -i --zerocounters --directory BUILD_COVERAGE/
Deleting all .da files in BUILD_COVERAGE/ and subdirectories
Done.
lcov --capture --initial --directory BUILD_COVERAGE -b BUILD_COVERAGE --output-file coverage/broker.init.info
Capturing coverage data from BUILD_COVERAGE
Can't locate Digest/MD5.pm in @INC (@INC contains: /usr/local/lib64/perl5 /usr/local/share/perl5 /usr/lib64/perl5/vendor_perl /usr/share/perl5/vendor_perl /usr/lib64/perl5 /usr/share/perl5 .) at /usr/bin/geninfo line 57.
BEGIN failed--compilation aborted at /usr/bin/geninfo line 57. 
make: *** [coverage] Error 2
```
- libxslt
```
        1119 - 3389_forced_update_option_bool (Failed)
        1120 - 3389_forced_update_option_number (Failed)
        1121 - 3389_forced_update_option_string (Failed)
        1132 - legacy-forwarding-with-idpattern (Failed)
Errors while running CTest
make[1]: *** [test] Error 8
make[1]: Leaving directory `/opt/fiware-orion/BUILD_COVERAGE'
/bin/sh: xsltproc: command not found
make: *** [coverage] Error 127 
```

fgalan replied:
Ok. NTC

fgalan commented on doc/manuals/admin/build_source.md:
```diff
@@ -85,16 +86,29 @@ for details. Recommended version is 3.6 (although 3.2 and 3.4 should also work f
 * Install the required tools:
 
-        sudo yum install python python-flask pyOpenSSL curl nc mongodb-org-shell valgrind bc
+        sudo yum install python curl nc mongodb-org-shell valgrind bc python-devel libffi-devel python-pip
```
I understand your approach and I think is better to use pip dependencies (installed in a virtualenv) instead that system modules from python-flask and pyOpenSSL. Thus, I understand the need for python-devel and python-pip. However, what about libffi-devel? What's it? Could you elaborate, please?

fisuda replied:
The pyOpenSSL requires the cryptography as a dependncy.
In the case of x86_64, the crypotgrapy is installed from a .whl file. 
```
Collecting cryptography>=2.8
  Downloading cryptography-2.8-cp27-cp27mu-manylinux2010_x86_64.whl (2.3 MB) 
```
In the case of aarch64, the crypotgrapy is built from a source code. The libffi-devel is needed as a dependcy library for build.
```
Collecting cryptography>=2.8
  Downloading cryptography-2.8.tar.gz (504 kB) 
```

The details are the following:
```
Collecting cryptography>=2.8
  Using cached cryptography-2.8.tar.gz (504 kB)
  Installing build dependencies ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python2 /usr/lib/python2.7/site-packages/pip install --ignore-installed --no-user --prefix /tmp/pip-build-env-aYE2lL/overlay --no-warn-script-location --no-binary :none: --only-binary :none: -i https://pypi.org/simple -- 'setuptools>=40.6.0' wheel 'cffi>=1.8,!=1.11.3; platform_python_implementation != '"'"'PyPy'"'"''
       cwd: None
  Complete output (88 lines):
  DEPRECATION: Python 2.7 reached the end of its life on January 1st, 2020. Please upgrade your Python as Python 2.7 is no longer maintained. A future version of pip will drop support for Python 2.7. More details about Python 2 support in pip, can be found at https://pip.pypa.io/en/latest/development/release-process/#python-2-support
  Collecting setuptools>=40.6.0
    Using cached setuptools-44.0.0-py2.py3-none-any.whl (583 kB)
  Collecting wheel
    Using cached wheel-0.34.2-py2.py3-none-any.whl (26 kB)
  Collecting cffi!=1.11.3,>=1.8
    Using cached cffi-1.14.0.tar.gz (463 kB)
  Collecting pycparser
    Using cached pycparser-2.19.tar.gz (158 kB)
  Installing collected packages: setuptools, wheel, pycparser, cffi
      Running setup.py install for pycparser: started
      Running setup.py install for pycparser: finished with status 'done'
      Running setup.py install for cffi: started
      Running setup.py install for cffi: finished with status 'error'
      ERROR: Command errored out with exit status 1:
       command: /usr/bin/python2 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-1i9FKC/cffi/setup.py'"'"'; __file__='"'"'/tmp/pip-install-1i9FKC/cffi/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-iJbp8Y/install-record.txt --single-version-externally-managed --prefix /tmp/pip-build-env-aYE2lL/overlay --compile --install-headers /tmp/pip-build-env-aYE2lL/overlay/include/python2.7/cffi
           cwd: /tmp/pip-install-1i9FKC/cffi/
      Complete output (68 lines):
      Package libffi was not found in the pkg-config search path.
      Perhaps you should add the directory containing `libffi.pc'
      to the PKG_CONFIG_PATH environment variable
      No package 'libffi' found
      Package libffi was not found in the pkg-config search path.
      Perhaps you should add the directory containing `libffi.pc'
      to the PKG_CONFIG_PATH environment variable
      No package 'libffi' found
      Package libffi was not found in the pkg-config search path.
      Perhaps you should add the directory containing `libffi.pc'
      to the PKG_CONFIG_PATH environment variable
      No package 'libffi' found
      Package libffi was not found in the pkg-config search path.
      Perhaps you should add the directory containing `libffi.pc'
      to the PKG_CONFIG_PATH environment variable
      No package 'libffi' found
      Package libffi was not found in the pkg-config search path.
      Perhaps you should add the directory containing `libffi.pc'
      to the PKG_CONFIG_PATH environment variable
      No package 'libffi' found
      unable to execute gcc: No such file or directory
      unable to execute gcc: No such file or directory
 ```

fgalan replied:
Thanks for the clarification!

I see this part has changed in the PR. I'll provide additional feedback on the new version. NTC with regards this comment thread.

fgalan commented on docker/raspberry_pi.md:
```diff
@@ -0,0 +1,94 @@
+# How to use Orion Context Broker on Raspberry Pi.
```
So they idea is to install docker in the Rasberry PI operating system, then build and run Orion as docker container on that system, isn't it? (vs. build and run Orion directly in the Raspberry PI operating system)

Just to check if I understand correctly :)

fisuda replied:
Your understanding is correct. It build and run Orion as docker container on a Raspberry Pi.

fgalan replied:
Ok. Thanks for the clarification!

NTC

fgalan commented:
Thanks for your contribution!

This covers a very old issue (which was closed due to oldness :) about running Orion on a Raspberry: #15. Great!

fisuda commented:
Thank you for your comments!

I updated English and Japanese build_source.md files.

I added the instraction to upgrade pip. In both cases of x86_64 and aarch64, if pip is not upgraded, The `ImportError` is occured when installing virtualenv.
```
ImportError: No module named zipp
```
As it was not necessary to specify the version of enum34, I deleted `enum34==1.1.6`. It is a dependency module of pyOpenSSL.

It would be great if you could review this update.

Thank you.

fgalan commented on docker/Dockerfile:
```diff
@@ -30,6 +36,7 @@ RUN \
       which \
       cyrus-sasl-devel && \
     # Install libmicrohttpd from source
+    cd /opt && \
```
Why the added `cd /opt &&` steps? Is some kind of fixing in the Dockerfile?

fisuda replied:
The directrices of the source codes are nested if the 'cd /opt' doesn't be added.
The source code of the Orion is as shown:
```
/opt/libmicrohttpd-0.9.48/mongo-cxx-driver-legacy-1.1.2/fiware-orion
```

fgalan replied:
Ok. Thanks for the fix!

NTC

fgalan commented on doc/manuals/admin/build_source.md:
```diff
@@ -85,16 +87,32 @@ for details. Recommended version is 3.6 (although 3.2 and 3.4 should also work f
+        gpgcheck=1
+        enabled=1
+        gpgkey=https://www.mongodb.org/static/pgp/server-3.6.asc
+
```
Apart from the mongodb.repo, maybe the yum command to actually install the mongo shell should be added?

fisuda replied:
Sorry, I made an editing mistake. I added 'mongodb-org-shell' in line 101 again.

fgalan replied:
Fixed in 6039fe1

fgalan commented on doc/manuals/admin/build_source.md:
```diff
@@ -85,16 +87,32 @@ for details. Recommended version is 3.6 (although 3.2 and 3.4 should also work f
 * Install the required tools:
 
-        sudo yum install python python-flask pyOpenSSL curl nc mongodb-org-shell valgrind bc
+        sudo yum install python python-devel python-pip curl nc valgrind bc
```
Looking to https://github.com/telefonicaid/fiware-orion/blob/master/ci/rpm7/build-dep.sh (which in some sense is "paralell" to the procedure described in build_source.md) I think `python-devel` is not needed here. 

Maybe `python-devel` is needed only for `libffi-devel`? In that case, "move" it to line 104 which should read as follows:

> In the case of the aarch64 architecture, additionally install libffi-devel and python-devel using yum. It is needed when building pyOpenSSL

fisuda replied:
Your comment is correct. The python-devel is needed only for the libffi-devel. I edited line 100 and 104.

fgalan replied:
Fixed in 6039fe1

fisuda commented:
Thank you very much for merging it!! :smile:
